### ISSUE TITLE
How to keep the current optimized backward that does not support double backward

### ISSUE THREAD

**Issue by beam2d** [chainer/chainer#3146]
In order to support the new-style function (#2970), we have to reimplement `backward` of each function node implementation that supports double backward. The easiest way is to implement `backward` using other differentiable functions so that it automatically supports double backward. However, this naive approach does not work well with the current optimized backward implementations because they use numpy/cupy level functions, esp. fused kernels of cupy, which cannot be differentiated automatically.

There are at least two approaches to keep the performance yet obtain the ability to backprop twice.

1. Switch the fast (single backward) and slow (double backward) implementations at runtime based on user requirements. See the comment by @jekbradbury  https://github.com/chainer/chainer/pull/3119#issuecomment-320759369. For example, add `differentaible` option to `Variable.backward()` which is `False` by default. If the flag is `False`, it uses the conventional implementations. Otherwise, it uses implementations that support double backward.
2. Implement each `backward` as a brand new function node whose forward is the current optimized code. In this case, we have to derive the second order gradient by ourselves to implement the `backward` of this new function. In this case, the performance is kept even if double backward is used.

Note that both cases need at least one new backward implementation at some point (either the first order grad or the second order grad). 1 is cheaper because we do not need to derive the second order derivative by ourselves. 2 less involves an additional API design. Which approach is preferred?

(I would like to decide in a couple of days so that we have enough time to work on the first-round implementations of new-style functions for v3)

**Comment by beam2d** [comment]
I have had an off-line discussion with @unnonouno. Here is the summary. See #3147 for what we want to support in the first round.

- Within functions listed in #3147, the difficult ones are `convolution_2d`, `*_pooling_2d`, `softmax_cross_entropy`, `lstm`, and `batch_normalization`.
- `convolution_2d` and `*_pooling_2d` require the second order grad formula written down by hand anyway because we currently do not have such functions.
- The difficulty of `softmax_cross_entropy` comes from its specification; it has many options that change the behavior (i.e., it is like a set of many different functions). This complication is orthogonal to the choice of the above two approaches.
- For `lstm` and `batch_normalization`, the approach 1 would be easier.

I think other functions are so simple that both approaches should work, so we should focus on `lstm` and `batch_normalization` for the current discussions. I feel adding another change to the core API is an overkill for supporting only two functions smoothly.

Based on these discussions, my opinion is to take the approach 2 at least in the first round. If implementing the second order grad is prohibitively costly, let us think again.

**Comment by beam2d** [comment]
Most of the above discussions are already obsolete since many of the individual functions now support double backprop. One concern is when users write their own functions and want to switch the implementation of `backward` for single-backprop usage and double-backprop usage. It looks we can look at `chainer.config.enable_backprop` flag to determine which value is passed to `enable_double_backprop` argument of `Variable.backward` or `chainer.grad`, so users can also use it. The problem is that it is not documented, so I think it's better to be clearly documented.
